WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,114 @@
+                                                                 onboardingPresenter: eventAdaptor,
+                                                                 configuration: CardPresentConfigurationLoader().configuration,
+                                                                 alertsPresenter: eventAdaptor)
+            orderPaymentUseCase.collectPayment(using: discoveryMethod) { error in
+                if let error = error as? CardPaymentErrorProtocol {
+                    streamContinuation.yield(.failure(error))
+                } else {
+                    streamContinuation.yield(.failure(CardPaymentsAdaptorError.unknownPaymentError(underlyingError: error)))
+                    streamContinuation.finish()
+                }
+            } onCancel: {
+                streamContinuation.yield(.cancellation)
+                streamContinuation.finish()
+            } onPaymentCompletion: {
+                // no-op – not used in PaymentMethodsViewModel anyway so this can be removed
+            } onCompleted: {
+                streamContinuation.yield(.success(order))
+                streamContinuation.finish()
+            }
+        }
+
+        return eventStream
+    }
+
+    enum CardPaymentsAdaptorError: Error, CardPaymentErrorProtocol {
+        var retryApproach: CardPaymentRetryApproach {
+            .restart
+        }
+
+        case unknownPaymentError(underlyingError: Error)
+    }
+}
+
+class CardPresentPaymentsEventAdaptor: CardPresentPaymentsOnboardingPresenting, CardPresentPaymentAlertsPresenting {
+    let eventStreamContinuation: AsyncStream<CardPresentPaymentEvent>.Continuation
+
+    init(eventStreamContinuation: AsyncStream<CardPresentPaymentEvent>.Continuation) {
+        self.eventStreamContinuation = eventStreamContinuation
+    }
+
+    func showOnboardingIfRequired(from: UIViewController, readyToCollectPayment: @escaping () -> Void) {
+        eventStreamContinuation.yield(.showOnboarding)
+    }
+
+    func refresh() {
+        // TODO: Refresh onboarding
+    }
+
+    func present(viewModel: CardPresentPaymentsModalViewModel) {
+        eventStreamContinuation.yield(.presentAlert(CardPresentPaymentsAdaptorPaymentAlert(from: viewModel)))
+    }
+
+    func foundSeveralReaders(readerIDs: [String], connect: @escaping (String) -> Void, cancelSearch: @escaping () -> Void) {
+        eventStreamContinuation.yield(.presentReaderList(readerIDs))
+    }
+
+    func updateSeveralReadersList(readerIDs: [String]) {
+        eventStreamContinuation.yield(.presentReaderList(readerIDs))
+    }
+
+    func dismiss() {
+        eventStreamContinuation.yield(.dismissAlerts)
+        // I don't know whether we _really_ need to implement this.
+        // It seems better to dismiss implicitly (if we need to present something else) or if the user triggers it.
+    }
+}
+
+enum CardPresentPaymentEvent {
```

Comment:
My guess is that this is for the consumer of this adaptor to have any UI implementation?

Reply:
Right now, yes. 

I'm going to experiment with adapting the existing UI to be presentable without compromises in SwiftUI, but I'm not particularly hopeful about that.